#VRML_SIM R2023b utf8

EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/backgrounds/protos/TexturedBackground.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/backgrounds/protos/TexturedBackgroundLight.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/floors/protos/Floor.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/appearances/protos/SandyGround.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/road/protos/Road.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/appearances/protos/Pavement.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/buildings/protos/Windmill.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/buildings/protos/SmallManor.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/factory/manhole/protos/SquareManhole.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/factory/containers/protos/CardboardBox.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/vehicles/protos/tesla/TeslaModel3Simple.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/trees/protos/Pine.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/trees/protos/Forest.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/robots/dji/mavic/protos/Mavic2Pro.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/geometries/protos/TexturedParallelepiped.proto"

WorldInfo {
  info [
    "This demonstration shows a DJI Mavic 2 PRO flying over a rural area."
    "After the drone has taken off, you can control it with the keyboard of your computer."
  ]
  title "DJI Mavic 2 PRO demo"
  basicTimeStep 8
  defaultDamping Damping {
    linear 0.5
    angular 0.5
  }
}
Viewpoint {
  orientation -0.22246638423549023 0.6921511623784 0.6866115905687995 0.5878111116242991
  position -1.6670014246003038 -0.8305502514399336 0.6132853114863441
  near 0.2
  follow "Mavic 2 PRO"
  followType "None"
  followSmoothness 0.2
}
TexturedBackground {
  luminosity 3
}
TexturedBackgroundLight {
}
Floor {
  size 400 400
  tileSize 6 6
  appearance SandyGround {
    colorOverride 0 0 0
  }
}
Solid {
  translation -0.9298433772447465 0.38 0.04990189999999999
  rotation -0.049805727667186245 0.03984013708934723 0.9979640038439483 3.997318339782539e-17
  children [
    Shape {
      geometry Box {
        size 0.1 0.1 0.1
      }
    }
  ]
  boundingObject Box {
    size 0.1 0.1 0.1
  }
  physics Physics {
  }
}
Solid {
  translation -1.52 0.38 0.04990189999999999
  rotation 0.1376594107515851 -0.30466052691094053 -0.942460529663643 7.672732755289756e-18
  children [
    Shape {
      geometry Box {
        size 0.1 0.1 0.1
      }
    }
  ]
  name "solid(1)"
  boundingObject Box {
    size 0.1 0.1 0.1
  }
  physics Physics {
  }
}
Solid {
  translation -1.26 0.1 0.04990189999999999
  rotation -0.02683864517255618 -0.0029915837500242345 0.9996353022737683 1.049464987150002e-16
  children [
    Shape {
      geometry Box {
        size 0.1 0.1 0.1
      }
    }
  ]
  name "solid(2)"
  boundingObject Box {
    size 0.1 0.1 0.1
  }
  physics Physics {
  }
}
Solid {
  translation 0.2 2.18 0.0499019
  rotation 0.6380917252399116 0.2987540852158757 -0.7096371937456342 2.1245406121402202e-17
  children [
    Shape {
      geometry Box {
        size 0.1 0.1 0.1
      }
    }
  ]
  name "solid(3)"
  boundingObject Box {
    size 0.1 0.1 0.1
  }
  physics Physics {
  }
}
Solid {
  translation -2 0.65 0.04990189999999999
  rotation 0.018430980184664193 -0.24432882789904467 0.9695172627792201 3.2355533945674963e-17
  children [
    Shape {
      geometry Box {
        size 0.1 0.1 0.1
      }
    }
  ]
  name "solid(4)"
  boundingObject Box {
    size 0.1 0.1 0.1
  }
  physics Physics {
  }
}
Solid {
  translation -1.26 0.66 0.04990189999999999
  rotation 0.17577324103469708 0.5962777933669037 0.7832985132589326 4.1826053390174255e-18
  children [
    Shape {
      geometry Box {
        size 0.1 0.1 0.1
      }
    }
  ]
  name "solid(5)"
  boundingObject Box {
    size 0.1 0.1 0.1
  }
  physics Physics {
  }
}
Solid {
  translation -2.12 0.25 0.04990189999999999
  rotation -0.027786260992194556 -0.10891426848443152 0.9936627223663846 4.950477164109314e-17
  children [
    Shape {
      geometry Box {
        size 0.1 0.1 0.1
      }
    }
  ]
  name "solid(6)"
  boundingObject Box {
    size 0.1 0.1 0.1
  }
  physics Physics {
  }
}
Solid {
  translation 0.74 2.03 0.0499019
  rotation -0.27876441232438465 0.045327781951488955 0.9592892132223726 6.809598057460623e-17
  children [
    Shape {
      geometry Box {
        size 0.1 0.1 0.1
      }
    }
  ]
  name "solid(7)"
  boundingObject Box {
    size 0.1 0.1 0.1
  }
  physics Physics {
  }
}
Solid {
  translation -2.68 1.51 0.049901899999999985
  rotation -0.0024234113342304583 -0.29815323784574316 0.9545149416534059 5.720080496590875e-17
  children [
    Shape {
      geometry Box {
        size 0.1 0.1 0.1
      }
    }
  ]
  name "solid(8)"
  boundingObject Box {
    size 0.1 0.1 0.1
  }
  physics Physics {
  }
}
Solid {
  translation 0.53 1.42 0.0499019
  rotation 0.1771611787371879 -0.044332421836172996 0.9831828686071528 2.2842401849924707e-17
  children [
    Shape {
      geometry Box {
        size 0.1 0.1 0.1
      }
    }
  ]
  name "solid(9)"
  boundingObject Box {
    size 0.1 0.1 0.1
  }
  physics Physics {
  }
}
Solid {
  translation -3.22 1.08 0.049901899999999985
  rotation -0.0024234113342304583 -0.29815323784574316 0.9545149416534059 5.720080496590875e-17
  children [
    Shape {
      geometry Box {
        size 0.1 0.1 0.1
      }
    }
  ]
  name "solid(10)"
  boundingObject Box {
    size 0.1 0.1 0.1
  }
  physics Physics {
  }
}
Solid {
  translation 0.16 1.31 0.0499019
  rotation 0.027599140319697853 -0.04608268379226784 0.9985562947115778 6.931323266997341e-17
  children [
    Shape {
      geometry Box {
        size 0.1 0.1 0.1
      }
    }
  ]
  name "solid(11)"
  boundingObject Box {
    size 0.1 0.1 0.1
  }
  physics Physics {
  }
}
Solid {
  translation -0.13 2.49 0.0499019
  rotation 0.6040598463408362 0.7849853874238818 -0.13751233969961926 4.4066144001503346e-18
  children [
    Shape {
      geometry Box {
        size 0.1 0.1 0.1
      }
    }
  ]
  name "solid(12)"
  boundingObject Box {
    size 0.1 0.1 0.1
  }
  physics Physics {
  }
}
Solid {
  translation -0.11 1.9 0.0499019
  rotation 0.02420947017321055 0.0843994966461879 0.9961378551785904 4.870028590679476e-17
  children [
    Shape {
      geometry Box {
        size 0.1 0.1 0.1
      }
    }
  ]
  name "solid(13)"
  boundingObject Box {
    size 0.1 0.1 0.1
  }
  physics Physics {
  }
}
Solid {
  translation -5.2 0.5 0.0499019
  rotation 0.04064398553516264 0.1680865060475383 0.9849340043498084 2.9964273852468753e-17
  children [
    Shape {
      geometry Box {
        size 0.1 0.1 0.1
      }
    }
  ]
  name "solid(14)"
  boundingObject Box {
    size 0.1 0.1 0.1
  }
  physics Physics {
  }
}
Solid {
  translation -2.75 2.47 0.04990189999999999
  rotation 0.18527191007727034 0.055969268981386 0.9810921262888666 7.952736166322876e-17
  children [
    Shape {
      geometry Box {
        size 0.1 0.1 0.1
      }
    }
  ]
  name "solid(15)"
  boundingObject Box {
    size 0.1 0.1 0.1
  }
  physics Physics {
  }
}
Road {
  translation -43.8 9.2 0.01
  rotation 0 0 1 -0.3926996938995747
  width 3
  numberOfLanes 1
  rightBorder FALSE
  leftBorder FALSE
  wayPoints [
    0 0 0
    100 0 0
    200 200 0
  ]
  splineSubdivision 8
  appearance Pavement {
    type "black stone"
  }
}
Windmill {
  translation -43.8656 -19.8433 0
  rotation -7.19228e-09 2.6842e-08 1 0.523601
}
Windmill {
  translation -44.258 -27.3356 0
  rotation -7.19228e-09 2.6842e-08 1 0.523601
  name "windmill(1)"
}
Windmill {
  translation -46.2448 30.573 0
  rotation -7.19228e-09 2.6842e-08 1 0.523601
  name "windmill(2)"
}
Windmill {
  translation -38.75 23.4 0
  name "windmill(3)"
}
SmallManor {
  translation -50.35 11.25 0
  rotation 0 0 -1 0.392394
}
SquareManhole {
  translation 0 0 -0.03
}
CardboardBox {
  translation -0.730157 -1.22891 0.3
  rotation 0 0 1 -1.8325953071795862
}
TeslaModel3Simple {
  translation -41.5139 4.34169 0.31
  rotation 0 0 1 -0.2618053071795865
}
Pine {
  translation -14.0083 -14.4828 0
  rotation 0 0 1 2.61799
}
Pine {
  translation -19.6195 -24.3783 0
  rotation 0 0 1 2.87979
  name "pine tree(1)"
}
Pine {
  translation -22.0129 6.20205 0
  name "pine tree(3)"
}
Pine {
  translation -10.6926 -25.1339 0
  name "pine tree(4)"
}
Pine {
  translation -14.3441 14.5621 0
  name "pine tree(5)"
}
Pine {
  translation -26.6328 -7.1707 0
  name "pine tree(6)"
}
Pine {
  translation -9.37029 14.0209 0
  rotation 0 0 1 -2.6179953071795863
  name "pine tree(2)"
}
Forest {
  translation 0 0 -0.01
  shape [
    200 200
    200 0
    50 0
    50 50
    -50 50
    -50 -50
    50 -50
    50 0
    200 0
    200 -200
    -200 -200
    -200 200
  ]
  density 0.01
  type "cherry tree"
  randomSeed 1
  maxHeight 12
  minHeight 6
  maxRadius 6
  minRadius 3
}
Mavic2Pro {
  translation -0.080758 -0.0439821 0.175492
  rotation 0.03485615736840446 -0.02158160516782757 0.9991592879075333 2.033532744508607
  bodySlot [
    Lidar {
      translation -0.13 0 -0.05
      rotation 0.5773489358556708 0.5773509358554485 -0.5773509358554485 2.0944
      children [
        Shape {
          geometry Box {
            size 0.01 0.01 0.01
          }
        }
      ]
      name "lidar(1)"
      horizontalResolution 20
      fieldOfView 1.2
      verticalFieldOfView 1.2
      numberOfLayers 10
      maxRange 2
    }
    Lidar {
      translation -0.11 0 -0.05
      rotation 0 1 0 1.309
      children [
        Shape {
          geometry Box {
            size 0.01 0.01 0.01
          }
        }
      ]
      boundingObject Box {
        size 0.01 0.01 0.01
      }
      horizontalResolution 20
      fieldOfView 1.2
      verticalFieldOfView 0.4
      numberOfLayers 7
      maxRange 2
    }
    DistanceSensor {
      translation -0.01 0 -0.06
      rotation 0 1 0 1.5708
      children [
        Shape {
          geometry Box {
            size 0.01 0.01 0.01
          }
        }
      ]
      name "distance"
      boundingObject Box {
        size 0.01 0.01 0.01
      }
      lookupTable [
        0 0 0
        0.9 1000 0
      ]
    }
    GPS {
      translation -0.1 0 0
      name "gps(2)"
    }
    GPS {
      translation 0.07 0 0
      name "gps(1)"
    }
    HingeJoint {
      jointParameters HingeJointParameters {
        axis 0 -1 0
        anchor -0.0725 0 -0.107
      }
      device [
        PositionSensor {
          name "PS2"
        }
        RotationalMotor {
          name "RM2"
          acceleration 1
        }
      ]
      endPoint Solid {
        translation -0.0725 0 -0.107
        rotation 2.858809999979949e-06 2.419629999983029e-06 -0.9999999999929862 1.5707953071795862
        children [
          Solid {
            translation 0 -0.027 -0.0193
            rotation -1.14927e-06 1.40791e-07 -1 -3.1415853071795863
            children [
              DEF claw Shape {
                appearance PBRAppearance {
                  baseColor 0 1 0
                  roughness 1
                  metalness 0
                }
                geometry TexturedParallelepiped {
                  size 0.005 0.01 0.04
                  angles 0 1 0
                }
              }
            ]
            name "solid(1)"
            boundingObject USE claw
            physics Physics {
            }
          }
          Solid {
            translation 0 -0.0225 0
            rotation -1 6.86169e-07 2.69628e-07 1.57079
            children [
              DEF arm Shape {
                appearance PBRAppearance {
                  baseColor 0 1 0
                  roughness 1
                  metalness 0
                }
                geometry Box {
                  size 0.005 0.005 0.04
                }
              }
            ]
            boundingObject USE arm
            physics Physics {
            }
          }
          DEF axle Shape {
            appearance PBRAppearance {
              baseColor 1 0 1
              roughness 1
              metalness 0
            }
            geometry Box {
              size 0.005 0.005 0.005
            }
          }
        ]
        name "solid(2)"
        contactMaterial "solid"
        boundingObject USE axle
        physics Physics {
        }
      }
    }
    HingeJoint {
      jointParameters HingeJointParameters {
        axis -1 0 0
        anchor -0.06 0.0125 -0.107
      }
      device [
        PositionSensor {
          name "PS3"
        }
        RotationalMotor {
          name "RM3"
          acceleration 1
        }
      ]
      endPoint Solid {
        translation -0.06 0.0125 -0.107
        rotation -2.1958999999922994e-07 -2.639219999990745e-06 0.9999999999964932 3.14159
        children [
          Solid {
            translation 0 -0.027 -0.0193
            rotation -1.14927e-06 1.40791e-07 -1 -3.1415853071795863
            children [
              DEF claw Shape {
                appearance PBRAppearance {
                  baseColor 0 1 0
                  roughness 1
                  metalness 0
                }
                geometry TexturedParallelepiped {
                  size 0.005 0.01 0.04
                  angles 0 1 0
                }
              }
            ]
            name "solid(1)"
            boundingObject USE claw
            physics Physics {
            }
          }
          Solid {
            translation 0 -0.0225 0
            rotation -1 6.86169e-07 2.69628e-07 1.57079
            children [
              DEF arm Shape {
                appearance PBRAppearance {
                  baseColor 0 1 0
                  roughness 1
                  metalness 0
                }
                geometry Box {
                  size 0.005 0.005 0.04
                }
              }
            ]
            boundingObject USE arm
            physics Physics {
            }
          }
          DEF axle Shape {
            appearance PBRAppearance {
              baseColor 1 0 1
              roughness 1
              metalness 0
            }
            geometry Box {
              size 0.005 0.005 0.005
            }
          }
        ]
        name "solid(3)"
        contactMaterial "solid"
        boundingObject USE axle
        physics Physics {
          density 1
        }
      }
    }
    HingeJoint {
      jointParameters HingeJointParameters {
        anchor -0.06 -0.0125 -0.107
      }
      device [
        PositionSensor {
          name "PS1"
        }
        RotationalMotor {
          name "RM1"
          acceleration 1
        }
      ]
      endPoint Solid {
        translation -0.06 -0.0125 -0.107
        rotation 0.9945896626077083 -0.08275237192810922 -0.06279687869756383 5.307156870259904e-06
        children [
          Solid {
            translation 0 -0.027 -0.0193
            rotation -1.14927e-06 1.40791e-07 -1 -3.1415853071795863
            children [
              DEF claw Shape {
                appearance PBRAppearance {
                  baseColor 0 1 0
                  roughness 1
                  metalness 0
                }
                geometry TexturedParallelepiped {
                  size 0.005 0.01 0.04
                  angles 0 1 0
                }
              }
            ]
            name "solid(1)"
            boundingObject USE claw
            physics Physics {
            }
          }
          Solid {
            translation 0 -0.0225 0
            rotation -1 6.86169e-07 2.69628e-07 1.57079
            children [
              DEF arm Shape {
                appearance PBRAppearance {
                  baseColor 0 1 0
                  roughness 1
                  metalness 0
                }
                geometry Box {
                  size 0.005 0.005 0.04
                }
              }
            ]
            boundingObject USE arm
            physics Physics {
            }
          }
          DEF axle Shape {
            appearance PBRAppearance {
              baseColor 1 0 1
              roughness 1
              metalness 0
            }
            geometry Box {
              size 0.005 0.005 0.005
            }
          }
        ]
        name "solid(1)"
        contactMaterial "solid"
        boundingObject USE axle
        physics Physics {
        }
      }
    }
    Pose {
      translation -0.06 0 -0.107
      rotation 0 1 0 -5.307179586466759e-06
      children [
        DEF body Shape {
          appearance PBRAppearance {
            baseColor 0 1 1
            roughness 1
            metalness 0
          }
          geometry Box {
            size 0.02 0.02 0.005
          }
        }
      ]
    }
    HingeJoint {
      jointParameters HingeJointParameters {
        axis 0 1 0
        anchor -0.0475 0 -0.107
      }
      device [
        PositionSensor {
          name "PS"
        }
        RotationalMotor {
          name "RM"
          acceleration 1
        }
      ]
      endPoint Solid {
        translation -0.0475 0 -0.107
        rotation 1.5456099999968308e-06 1.3081699999973181e-06 -0.9999999999979499 1.570795307181636
        children [
          Solid {
            translation 0 0.027 -0.0193
            rotation 0.04429487720597257 0.3615768139335216 -0.9312895207608595 6.35691e-06
            children [
              DEF claw Shape {
                appearance PBRAppearance {
                  baseColor 0 1 0
                  roughness 1
                  metalness 0
                }
                geometry TexturedParallelepiped {
                  size 0.005 0.01 0.04
                  angles 0 1 0
                }
              }
            ]
            name "solid(1)"
            boundingObject USE claw
            physics Physics {
            }
          }
          Solid {
            translation 0 0.0225 0
            rotation 1 0 0 1.01503e-06
            children [
              DEF arm Shape {
                appearance PBRAppearance {
                  baseColor 0 1 0
                  roughness 1
                  metalness 0
                }
                geometry Box {
                  size 0.005 0.04 0.005
                }
              }
            ]
            boundingObject USE arm
            physics Physics {
            }
          }
          DEF axle Shape {
            appearance PBRAppearance {
              baseColor 1 0 1
              roughness 1
              metalness 0
            }
            geometry Box {
              size 0.005 0.005 0.005
            }
          }
        ]
        contactMaterial "solid"
        boundingObject USE axle
        physics Physics {
        }
      }
    }
    Pose {
      translation -0.06 0 -0.08
      children [
        DEF body Shape {
          appearance PBRAppearance {
            baseColor 0 1 1
            roughness 1
            metalness 0
          }
          geometry Box {
            size 0.005 0.005 0.05
          }
        }
      ]
    }
  ]
  cameraSlot [
    Camera {
      width 400
      height 240
      near 0.2
    }
  ]
}
